File: templates/python-crewai/src/tools.py
"""This module defines the tools used by the agent.

Feel free to modify or add new tools to suit your specific needs.

To learn how to create a new tool, see:
- https://docs.crewai.com/concepts/tools
"""

from __future__ import annotations

import os

from apify import Actor
from apify_client import ApifyClient
from crewai.tools import BaseTool
from crewai.utilities.converter import ValidationError
from pydantic import BaseModel, Field

from src.models import InstagramPost, InstagramPosts


class InstagramScraperInput(BaseModel):
    """Input schema for InstagramScraper tool."""

    handle: str = Field(..., description="Instagram handle of the profile to scrape (without the '@' symbol).")
    max_posts: int = Field(default=30, description='Maximum number of posts to scrape.')


class InstagramScraperTool(BaseTool):
    """Tool for scraping Instagram profile posts."""

    name: str = 'Instagram Profile Posts Scraper'
    description: str = 'Tool to scrape Instagram profile posts.'
    args_schema: type[BaseModel] = InstagramScraperInput

    def _run(self, handle: str, max_posts: int = 30) -> list[InstagramPost]:
        run_input = {
            'directUrls': [f'https://www.instagram.com/{handle}/'],
            'resultsLimit': max_posts,
            'resultsType': 'posts',
            'searchLimit': 1,
        }
        if not (token := os.getenv('APIFY_TOKEN')):
            raise ValueError('APIFY_TOKEN environment variable is missing!')

        apify_client = ApifyClient(token=token)
        if not (run := apify_client.actor('apify/instagram-scraper').call(run_input=run_input)):
            msg = 'Failed to start the Actor apify/instagram-scraper'
            raise RuntimeError(msg)

        dataset_id = run['defaultDatasetId']
        dataset_items: list[dict] = (apify_client.dataset(dataset_id).list_items()).items

        try:
            posts: InstagramPosts = InstagramPosts.model_validate(dataset_items)
        except ValidationError as e:
            Actor.log.warning('Received invalid dataset items: %s. Error: %s', dataset_items, e)
            raise RuntimeError('Received invalid dataset items.') from e
        else:
            return posts.root

File: templates/ts-mastraai/src/tools.ts
// eslint-disable-next-line import/extensions
import type { Tool } from '@mastra/core/tools';
// eslint-disable-next-line import/extensions
import { createTool } from '@mastra/core/tools';
import { ApifyClient, log } from 'apify';
import { z } from 'zod';

import { InstagramPosts, InstagramPostSchema } from './models.js';

const instagramScraperInputSchema = z.object({
    handle: z.string().describe("Instagram handle of the profile to scrape (without the '@' symbol)."),
    maxPosts: z.number().default(30).describe('Maximum number of posts to scrape.'),
});

const instagramScraperOutputSchema = z.object({
    posts: z.array(InstagramPostSchema).describe('List of scraped Instagram posts'),
});

// Define the Instagram Scraper Tool
export const instagramScraperTool: Tool<
    'instagram-profile-posts-scraper',
    typeof instagramScraperInputSchema,
    typeof instagramScraperOutputSchema
> = createTool({
    id: 'instagram-profile-posts-scraper',
    description: "Tool to scrape Instagram profile posts using Apify's Instagram Scraper.",
    inputSchema: instagramScraperInputSchema,
    outputSchema: instagramScraperOutputSchema,
    execute: async ({ context }) => {
        const token = process.env.APIFY_TOKEN;
        if (!token) {
            throw new Error('APIFY_TOKEN environment variable is missing!');
        }

        const { handle, maxPosts } = context;
        const runInput = {
            directUrls: [`https://www.instagram.com/${handle}/`],
            resultsLimit: maxPosts,
            resultsType: 'posts',
            searchLimit: 1,
        };

        const apifyClient = new ApifyClient({ token });

        // Call the Apify Instagram Scraper actor
        const run = await apifyClient.actor('apify/instagram-scraper').call(runInput);
        if (!run) {
            throw new Error('Failed to start the Actor apify/instagram-scraper');
        }

        // Fetch dataset items
        const datasetId = run.defaultDatasetId;
        const dataset = await apifyClient.dataset(datasetId).listItems();
        const datasetItems: unknown[] = dataset.items;

        // Validate and convert dataset items to InstagramPosts
        try {
            const posts: InstagramPosts = InstagramPosts.fromRaw(datasetItems);
            return { posts: posts.root };
        } catch (error) {
            log.warning(`Received invalid dataset items: ${JSON.stringify(datasetItems)}. Error: ${error}`);
            throw new Error('Received invalid dataset items.');
        }
    },
});
